### DISCUSSION TITLE
About disambiguation of .tsx files

### DISCUSSION THREAD

**Discussion by chibiricky** [github-linguist/linguist#6340] [General]
According to [this line of code in heuristics.yml](https://github.com/github/linguist/blob/master/lib/linguist/heuristics.yml#L719), in order for a .tsx file to be classified as language TSX (=> TypeScript), it has to have a line that imports the react module.

However, according to [this blog post that introduced React 17](https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#whats-a-jsx-transform), `With the new transform, you can use JSX without importing React`.

Also, [a JSX (or TSX) file does not have to be a React file](https://betterprogramming.pub/how-to-use-jsx-without-react-21d23346e5dc). I think we should not use the React import statement to identify TSX files.

Moreoever, when developing in next.js, there is even less chance that we need to import the React components, but the files are still TSX files.

Maybe in the .tsx section in the heuristics.yml file, we should put XML before TSX, and remove the pattern for TSX so that it can do a "catch-all" for all .tsx files that do not match the XML pattern?

**Comment by Alhadis** [comment]
Heuristics aren't a requirement (in most cases; see below[^1]). Here, they're being used to identify obvious TSX and XML files before passing the file down to Linguist's classifier strategy—which is the least accurate of Linguist's identification strategies (as well as the most computationally expensive).

In other words, `.tsx` files *don't* require the `react` module to be imported in order to be recognised as TSX.

[^1]: Some heuristics have an open-ended fallback (a `- language: …` entry without an accompanying `pattern:` field). This is analogous to the `else` branch of an `if … else if …` conditional.`return
